### ISSUE TITLE
Unable to build on command line (make)

### ISSUE THREAD

**Issue by hyproman** [PhilippC/keepass2android#2211]
I am unable to build from command line under Windows 10, with the following error across multiple projects:
`error MSB4057: The target "Restore" does not exist in the project.`

I have tried a few different versions of MSBuild and nuget without success.

Source tree head is here:
```
commit 9958a73d4a0daed2b1bc62a4f5d1391cefd30dbc (HEAD -> master, tag: 1.09e-r1,
upstream/master)
Author: Philipp Crocoll <philipp@crocoll.net>
Date:   Wed Feb 1 21:16:05 2023 +0100

    revert to targetSdkVersion 31 to avoid issues with notifications not showing
```

Various build tool versions:
```
>%MSBUILD% -version
Microsoft (R) Build Engine version 16.11.2+f32259642 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

16.11.2.50704

>java -version
openjdk version "11.0.2" 2019-01-15
OpenJDK Runtime Environment 18.9 (build 11.0.2+9)
OpenJDK 64-Bit Server VM 18.9 (build 11.0.2+9, mixed mode)

NuGet Version: 6.4.0.123

Android NDK Version: 25.2.9519653

(Using make.exe from Android NDK)
```

make --debug output:
```
C:\Users\me\source\forks\keepass2android>make --debug
MAKESHELL:
SHELL: cmd

MSBUILD: C:\Users\me\MSBuild\Current\Bin\MSBuild.exe

ANDROID_SDK_ROOT: C:\Users\me\AppData\Local\Android\Sdk\
ANDROID_HOME: C:\Users\me\AppData\Local\Android\Sdk\
ANDROID_NDK_ROOT: C:\Users\me\AppData\Local\Android\Sdk\ndk\25.2.9519653\
Makefile:121: Configuration environment variable not set.
Makefile:126: Flavor environment variable not set.
MSBUILD_PARAM:
nuget path: C:\Users\me\bin\nuget.exe

DEL stamp.nuget_*
Could Not Find C:\Users\me\source\forks\keepass2android\stamp.nuget_*
nuget restore src/KeePass.sln
MSBuild auto-detection: using msbuild version '16.11.2.50704' from 'C:\Users\me\MSBuild\Current\Bin'.
All packages listed in packages.config are already installed.
Restoring packages for C:\Users\me\source\forks\keepass2android\src\keepass2android\keepass2android-app.csproj...
Restoring packages for C:\Users\me\source\forks\keepass2android\src\Kp2aBusinessLogic\Kp2aBusinessLogic.csproj...
Generating MSBuild file C:\Users\me\source\forks\keepass2android\src\Kp2aBusinessLogic\obj\Kp2aBusinessLogic.csproj.nuget.g.props.
Generating MSBuild file C:\Users\me\source\forks\keepass2android\src\Kp2aBusinessLogic\obj\Kp2aBusinessLogic.csproj.nuget.g.targets.
Writing assets file to disk. Path: C:\Users\me\source\forks\keepass2android\src\Kp2aBusinessLogic\obj\project.assets.json
Restored C:\Users\me\source\forks\keepass2android\src\Kp2aBusinessLogic\Kp2aBusinessLogic.csproj (in 557 ms).
WARNING: NU1605: Detected package downgrade: Xamarin.AndroidX.AppCompat from 1.3.1.3 to 1.2.0.5. Reference the package directly from the project to select a different version.
 keepass2anrdoid-setup2 -> Xamarin.AndroidX.Biometric 1.1.0.6 -> Xamarin.AndroidX.AppCompat (>= 1.3.1.3)
 keepass2anrdoid-setup2 -> Xamarin.AndroidX.AppCompat (>= 1.2.0.5)
WARNING: NU1605: Detected package downgrade: Xamarin.AndroidX.Core from 1.6.0.3 to 1.5.0. Reference the package directly from the project to select a different version.
 keepass2anrdoid-setup2 -> Xamarin.AndroidX.Biometric 1.1.0.6 -> Xamarin.AndroidX.Core (>= 1.6.0.3)
 keepass2anrdoid-setup2 -> Xamarin.AndroidX.Core (>= 1.5.0)
WARNING: NU1605: Detected package downgrade: Xamarin.AndroidX.Core from 1.6.0.3 to 1.5.0. Reference the package directly from the project to select a different version.
 keepass2anrdoid-setup2 -> Xamarin.AndroidX.CustomView 1.1.0.9 -> Xamarin.AndroidX.Core (>= 1.6.0.3)
 keepass2anrdoid-setup2 -> Xamarin.AndroidX.Core (>= 1.5.0)
WARNING: NU1605: Detected package downgrade: Xamarin.AndroidX.AppCompat from 1.3.1.3 to 1.2.0.5. Reference the package directly from the project to select a different version.
 keepass2anrdoid-setup2 -> Xamarin.AndroidX.Preference 1.1.1.11 -> Xamarin.AndroidX.AppCompat (>= 1.3.1.3)
 keepass2anrdoid-setup2 -> Xamarin.AndroidX.AppCompat (>= 1.2.0.5)
WARNING: NU1605: Detected package downgrade: Xamarin.AndroidX.Core from 1.6.0.3 to 1.5.0. Reference the package directly from the project to select a different version.
 keepass2anrdoid-setup2 -> Xamarin.AndroidX.Preference 1.1.1.11 -> Xamarin.AndroidX.Core (>= 1.6.0.3)
 keepass2anrdoid-setup2 -> Xamarin.AndroidX.Core (>= 1.5.0)
WARNING: NU1605: Detected package downgrade: Xamarin.AndroidX.Core from 1.6.0.3 to 1.5.0. Reference the package directly from the project to select a different version.
 keepass2anrdoid-setup2 -> Xamarin.AndroidX.RecyclerView 1.2.1.3 -> Xamarin.AndroidX.Core (>= 1.6.0.3)
 keepass2anrdoid-setup2 -> Xamarin.AndroidX.Core (>= 1.5.0)
WARNING: NU1605: Detected package downgrade: Xamarin.AndroidX.AppCompat from 1.2.0.6 to 1.2.0.5. Reference the package directly from the project to select a different version.
 keepass2anrdoid-setup2 -> ZXing.Net.Mobile 3.1.0-beta2 -> Xamarin.AndroidX.AppCompat (>= 1.2.0.6)
 keepass2anrdoid-setup2 -> Xamarin.AndroidX.AppCompat (>= 1.2.0.5)
WARNING: NU1605: Detected package downgrade: Xamarin.AndroidX.AppCompat from 1.2.0.6 to 1.2.0.5. Reference the package directly from the project to select a different version.
 keepass2anrdoid-setup2 -> ZXing.Net.Mobile.Forms 3.1.0-beta2 -> Xamarin.AndroidX.AppCompat (>= 1.2.0.6)
 keepass2anrdoid-setup2 -> Xamarin.AndroidX.AppCompat (>= 1.2.0.5)
WARNING: NU1605: Detected package downgrade: Xamarin.AndroidX.Preference from 1.2.0.3 to 1.1.1.11. Reference the package directly from the project to select a different version.
 keepass2anrdoid-setup2 -> Kp2aBusinessLogic -> Xamarin.AndroidX.Preference (>= 1.2.0.3)
 keepass2anrdoid-setup2 -> Xamarin.AndroidX.Preference (>= 1.1.1.11)
WARNING: NU1605: Detected package downgrade: Xamarin.AndroidX.Lifecycle.LiveData from 2.2.0.4 to 2.2.0.3. Reference the package directly from the project to select a different version.
 keepass2anrdoid-setup2 -> ZXing.Net.Mobile.Forms 3.1.0-beta2 -> Xamarin.Forms 5.0.0.1874 -> Xamarin.AndroidX.Lifecycle.LiveData (>= 2.2.0.4)
 keepass2anrdoid-setup2 -> Xamarin.AndroidX.Lifecycle.LiveData (>= 2.2.0.3)
WARNING: NU1605: Detected package downgrade: Xamarin.Google.Android.Material from 1.1.0.5 to 1.0.0.1. Reference the package directly from the project to select a different version.
 keepass2anrdoid-setup2 -> ZXing.Net.Mobile.Forms 3.1.0-beta2 -> Xamarin.Forms 5.0.0.1874 -> Xamarin.Google.Android.Material (>= 1.1.0.5)
 keepass2anrdoid-setup2 -> Xamarin.Google.Android.Material (>= 1.0.0.1)
Generating MSBuild file C:\Users\me\source\forks\keepass2android\src\keepass2android\obj\keepass2android-app.csproj.nuget.g.props.
Generating MSBuild file C:\Users\me\source\forks\keepass2android\src\keepass2android\obj\keepass2android-app.csproj.nuget.g.targets.
Writing assets file to disk. Path: C:\Users\me\source\forks\keepass2android\src\keepass2android\obj\project.assets.json
Restored C:\Users\me\source\forks\keepass2android\src\keepass2android\keepass2android-app.csproj (in 970 ms).

NuGet Config files used:
    C:\Users\me\AppData\Roaming\NuGet\NuGet.Config
    C:\Program Files (x86)\NuGet\Config\Microsoft.VisualStudio.Offline.config
    C:\Program Files (x86)\NuGet\Config\Xamarin.Offline.config

Feeds used:
    C:\Program Files (x86)\Microsoft SDKs\NuGetPackages\
    https://api.nuget.org/v3/index.json
C:\Users\me\MSBuild\Current\Bin\MSBuild.exe src/KeePass.sln -t:restore  -p:RestorePackagesConfig=true
Microsoft (R) Build Engine version 16.11.2+f32259642 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

Building the projects in this solution one at a time. To enable parallel build, please add the "-m" switch.
Build started 2/2/2023 11:25:40 PM.
Project "C:\Users\me\source\forks\keepass2android\src\KeePass.sln" on node 1 (Restore target(s)).
ValidateSolutionConfiguration:
  Building solution configuration "Debug|Mixed Platforms".
Project "C:\Users\me\source\forks\keepass2android\src\KeePass.sln" (1) is building "C:\Users\me\source\forks\keepass2android\src\KeePassLib2An
droid\KeePassLib2Android.csproj" (2) on node 1 (Restore target(s)).
C:\Users\me\source\forks\keepass2android\src\KeePassLib2Android\KeePassLib2Android.csproj : error MSB4057: The target "Restore" does not exist in the project.
Done Building Project "C:\Users\me\source\forks\keepass2android\src\KeePassLib2Android\KeePassLib2Android.csproj" (Restore target(s)) -- FAILED.

Project "C:\Users\me\source\forks\keepass2android\src\KeePass.sln" (1) is building "C:\Users\me\source\forks\keepass2android\src\Kp2aKeyboardBinding\Kp2aKeyboardBinding.csproj" (3) on node 1 (Restore target(s)).
C:\Users\me\source\forks\keepass2android\src\Kp2aKeyboardBinding\Kp2aKeyboardBinding.csproj : error MSB4057: The target "Restore" does not exist in the project.
Done Building Project "C:\Users\me\source\forks\keepass2android\src\Kp2aKeyboardBinding\Kp2aKeyboardBinding.csproj" (Restore target(s)) -- FAILED.

Project "C:\Users\me\source\forks\keepass2android\src\KeePass.sln" (1) is building "C:\Users\me\source\forks\keepass2android\src\Kp2aBusinessLogic\Kp2aBusinessLogic.csproj" (4) on node 1 (Restore target(s)).
showFlavor:
  building flavor NoNet
C:\Users\me\source\forks\keepass2android\src\Kp2aBusinessLogic\Kp2aBusinessLogic.csproj : error MSB4057: The target "Restore" does not exist in the project.
Done Building Project "C:\Users\me\source\forks\keepass2android\src\Kp2aBusinessLogic\Kp2aBusinessLogic.csproj" (Restore target(s)) -- FAILED.

Project "C:\Users\me\source\forks\keepass2android\src\KeePass.sln" (1) is building "C:\Users\me\source\forks\keepass2android\src\TwofishCipher
\TwofishCipher.csproj" (5) on node 1 (Restore target(s)).
C:\Users\me\source\forks\keepass2android\src\TwofishCipher\TwofishCipher.csproj : error MSB4057: The target "Restore" does not exist in the project.
Done Building Project "C:\Users\me\source\forks\keepass2android\src\TwofishCipher\TwofishCipher.csproj" (Restore target(s)) -- FAILED.

Project "C:\Users\me\source\forks\keepass2android\src\KeePass.sln" (1) is building "C:\Users\me\source\forks\keepass2android\src\JavaFileStora
geBindings\JavaFileStorageBindings.csproj" (6) on node 1 (Restore target(s)).
C:\Users\me\source\forks\keepass2android\src\JavaFileStorageBindings\JavaFileStorageBindings.csproj : error MSB4057: The target "Restore" does not exist in the project.
Done Building Project "C:\Users\me\source\forks\keepass2android\src\JavaFileStorageBindings\JavaFileStorageBindings.csproj" (Restore target(s)) -- FAILED.

Project "C:\Users\me\source\forks\keepass2android\src\KeePass.sln" (1) is building "C:\Users\me\source\forks\keepass2android\src\AndroidFileCh
ooserBinding\AndroidFileChooserBinding.csproj" (7) on node 1 (Restore target(s)).
C:\Users\me\source\forks\keepass2android\src\AndroidFileChooserBinding\AndroidFileChooserBinding.csproj : error MSB4057: The target "Restore" does not exist in the project.
Done Building Project "C:\Users\me\source\forks\keepass2android\src\AndroidFileChooserBinding\AndroidFileChooserBinding.csproj" (Restore target(s)) -- FAILED.

Project "C:\Users\me\source\forks\keepass2android\src\KeePass.sln" (1) is building "C:\Users\me\source\forks\keepass2android\src\KP2AKdbLibrar
yBinding\KP2AKdbLibraryBinding.csproj" (8) on node 1 (Restore target(s)).
C:\Users\me\source\forks\keepass2android\src\KP2AKdbLibraryBinding\KP2AKdbLibraryBinding.csproj : error MSB4057: The target "Restore" does not exist in the project.
Done Building Project "C:\Users\me\source\forks\keepass2android\src\KP2AKdbLibraryBinding\KP2AKdbLibraryBinding.csproj" (Restore target(s)) -- FAILED.

Project "C:\Users\me\source\forks\keepass2android\src\KeePass.sln" (1) is building "C:\Users\me\source\forks\keepass2android\src\PluginSdkBind
ing\PluginSdkBinding.csproj" (9) on node 1 (Restore target(s)).
C:\Users\me\source\forks\keepass2android\src\PluginSdkBinding\PluginSdkBinding.csproj : error MSB4057: The target "Restore" does not exist in the project.
Done Building Project "C:\Users\me\source\forks\keepass2android\src\PluginSdkBinding\PluginSdkBinding.csproj" (Restore target(s)) -- FAILED.

Project "C:\Users\me\source\forks\keepass2android\src\KeePass.sln" (1) is building "C:\Users\me\source\forks\keepass2android\src\ZlibAndroid\Z
libAndroid.csproj" (10) on node 1 (Restore target(s)).
C:\Users\me\source\forks\keepass2android\src\ZlibAndroid\ZlibAndroid.csproj : error MSB4057: The target "Restore" does not exist in the project.
Done Building Project "C:\Users\me\source\forks\keepass2android\src\ZlibAndroid\ZlibAndroid.csproj" (Restore target(s)) -- FAILED.

Project "C:\Users\me\source\forks\keepass2android\src\KeePass.sln" (1) is building "C:\Users\me\source\forks\keepass2android\src\PCloudBinding
s\PCloudBindings.csproj" (11) on node 1 (Restore target(s)).
C:\Users\me\source\forks\keepass2android\src\PCloudBindings\PCloudBindings.csproj : error MSB4057: The target "Restore" does not exist in the project.
Done Building Project "C:\Users\me\source\forks\keepass2android\src\PCloudBindings\PCloudBindings.csproj" (Restore target(s)) -- FAILED.

Project "C:\Users\me\source\forks\keepass2android\src\KeePass.sln" (1) is building "C:\Users\me\source\forks\keepass2android\src\keepass2andro
id\keepass2android-app.csproj" (12) on node 1 (Restore target(s)).
showFlavor:
  building flavor NoNet
C:\Users\me\source\forks\keepass2android\src\keepass2android\keepass2android-app.csproj : error MSB4057: The target "Restore" does not exist in the project.
Done Building Project "C:\Users\me\source\forks\keepass2android\src\keepass2android\keepass2android-app.csproj" (Restore target(s)) -- FAILED.

Done Building Project "C:\Users\me\source\forks\keepass2android\src\KeePass.sln" (Restore target(s)) -- FAILED.


Build FAILED.

"C:\Users\me\source\forks\keepass2android\src\KeePass.sln" (Restore target) (1) ->
"C:\Users\me\source\forks\keepass2android\src\KeePassLib2Android\KeePassLib2Android.csproj" (Restore target) (2) ->
  C:\Users\me\source\forks\keepass2android\src\KeePassLib2Android\KeePassLib2Android.csproj : error MSB4057: The target "Restore" does not exist in the project.


"C:\Users\me\source\forks\keepass2android\src\KeePass.sln" (Restore target) (1) ->
"C:\Users\me\source\forks\keepass2android\src\Kp2aKeyboardBinding\Kp2aKeyboardBinding.csproj" (Restore target) (3) ->
  C:\Users\me\source\forks\keepass2android\src\Kp2aKeyboardBinding\Kp2aKeyboardBinding.csproj : error MSB4057: The target "Restore" does not exist in the project.


"C:\Users\me\source\forks\keepass2android\src\KeePass.sln" (Restore target) (1) ->
"C:\Users\me\source\forks\keepass2android\src\Kp2aBusinessLogic\Kp2aBusinessLogic.csproj" (Restore target) (4) ->
  C:\Users\me\source\forks\keepass2android\src\Kp2aBusinessLogic\Kp2aBusinessLogic.csproj : error MSB4057: The target "Restore" does not exist in the project.


"C:\Users\me\source\forks\keepass2android\src\KeePass.sln" (Restore target) (1) ->
"C:\Users\me\source\forks\keepass2android\src\TwofishCipher\TwofishCipher.csproj" (Restore target) (5) ->
  C:\Users\me\source\forks\keepass2android\src\TwofishCipher\TwofishCipher.csproj : error MSB4057: The target "Restore" does not exist in the project.


"C:\Users\me\source\forks\keepass2android\src\KeePass.sln" (Restore target) (1) ->
"C:\Users\me\source\forks\keepass2android\src\JavaFileStorageBindings\JavaFileStorageBindings.csproj" (Restore target) (6) ->
  C:\Users\me\source\forks\keepass2android\src\JavaFileStorageBindings\JavaFileStorageBindings.csproj : error MSB4057: The target "Restore" does not exist in the project.


"C:\Users\me\source\forks\keepass2android\src\KeePass.sln" (Restore target) (1) ->
"C:\Users\me\source\forks\keepass2android\src\AndroidFileChooserBinding\AndroidFileChooserBinding.csproj" (Restore target) (7) ->
  C:\Users\me\source\forks\keepass2android\src\AndroidFileChooserBinding\AndroidFileChooserBinding.csproj : error MSB4057: The target "Restore" does not exist in the project.


"C:\Users\me\source\forks\keepass2android\src\KeePass.sln" (Restore target) (1) ->
"C:\Users\me\source\forks\keepass2android\src\KP2AKdbLibraryBinding\KP2AKdbLibraryBinding.csproj" (Restore target) (8) ->
  C:\Users\me\source\forks\keepass2android\src\KP2AKdbLibraryBinding\KP2AKdbLibraryBinding.csproj : error MSB4057: The target "Restore" does not exist in the project.


"C:\Users\me\source\forks\keepass2android\src\KeePass.sln" (Restore target) (1) ->
"C:\Users\me\source\forks\keepass2android\src\PluginSdkBinding\PluginSdkBinding.csproj" (Restore target) (9) ->
  C:\Users\me\source\forks\keepass2android\src\PluginSdkBinding\PluginSdkBinding.csproj : error MSB4057: The target "Restore" does not exist in the project.


"C:\Users\me\source\forks\keepass2android\src\KeePass.sln" (Restore target) (1) ->
"C:\Users\me\source\forks\keepass2android\src\ZlibAndroid\ZlibAndroid.csproj" (Restore target) (10) ->
  C:\Users\me\source\forks\keepass2android\src\ZlibAndroid\ZlibAndroid.csproj : error MSB4057: The target "Restore" does not exist in the project.


"C:\Users\me\source\forks\keepass2android\src\KeePass.sln" (Restore target) (1) ->
"C:\Users\me\source\forks\keepass2android\src\PCloudBindings\PCloudBindings.csproj" (Restore target) (11) ->
  C:\Users\me\source\forks\keepass2android\src\PCloudBindings\PCloudBindings.csproj : error MSB4057: The target "Restore" does not exist in the project.


"C:\Users\me\source\forks\keepass2android\src\KeePass.sln" (Restore target) (1) ->
"C:\Users\me\source\forks\keepass2android\src\keepass2android\keepass2android-app.csproj" (Restore target) (12) ->
  C:\Users\me\source\forks\keepass2android\src\keepass2android\keepass2android-app.csproj : error MSB4057: The target "Restore" does not exist in the project.

    0 Warning(s)
    11 Error(s)

Time Elapsed 00:00:00.54
make: *** [Makefile:281: stamp.nuget_] Error 1
```

**Comment by tenzap** [comment]
Do you have visual studio installed? Which version?

**Comment by hyproman** [comment]
When I originally filed this bug, I had Visual Studio 2019 installed (somehow Googling Visual Studio with Xamarin support sent me to a page with the 2019 installer instead of 2022).

I have since completely cleared out and rebuilt my environment with Visual Studio 2022. Still can't get `make` to properly build.

**Problem 1: make doesn't like that MSBuild is located within a directory tree containing spaces**
```
C:\Users\me\source\forks\keepass2android>make --debug
MAKESHELL:
SHELL: cmd

MSBUILD: C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Current\Bin\MSBuild.exe

ANDROID_SDK_ROOT: C:\Users\me\AppData\Local\Android\Sdk\
ANDROID_HOME: C:\Users\me\AppData\Local\Android\Sdk\
ANDROID_NDK_ROOT: C:\Users\me\AppData\Local\Android\Sdk\ndk\25.2.9519653\
makefile:121: Configuration environment variable not set.
makefile:126: Flavor environment variable not set.
MSBUILD_PARAM:
nuget path: C:\Users\me\bin\nuget.exe

C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Current\Bin\MSBuild.exe src/KeePass.sln -target:keepass2android-app -p:AndroidSdkDirectory="C:\Users\me\AppData\Local\Android\Sdk\" -p:BuildProjectReferences=true  -p:Platform="Any CPU" -m
'C:\Program' is not recognized as an internal or external command,
operable program or batch file.
make: *** [msbuild] Error 1
```
So I did the `mklink /d` trick mentioned in the doc for Android SDK/NDK and linked `C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild` to `C:\Apps\MSBuild\vc`. I also added `C:\Apps\MSBuild\vc\Current\Bin` to the PATH so `make` sees it. This produces a different error:

**Problem 2: MSBuild doesn't handle the double-quotes around the `AndroidSdkDirectory` value properly**
```
C:\Users\me\source\forks\keepass2android>make --debug
MAKESHELL:
SHELL: cmd

MSBUILD: C:\Apps\MSBuild\vc\Current\Bin\MSBuild.exe

ANDROID_SDK_ROOT: C:\Users\me\AppData\Local\Android\Sdk\
ANDROID_HOME: C:\Users\me\AppData\Local\Android\Sdk\
ANDROID_NDK_ROOT: C:\Users\me\AppData\Local\Android\Sdk\ndk\25.2.9519653\
makefile:121: Configuration environment variable not set.
makefile:126: Flavor environment variable not set.
MSBUILD_PARAM:
nuget path: C:\Users\me\bin\nuget.exe

C:\Apps\MSBuild\vc\Current\Bin\MSBuild.exe src/KeePass.sln -target:keepass2android-app -p:AndroidSdkDirectory="C:\Users\me\AppData\Local\Android\Sdk\" -p:BuildProjectReferences=true  -p:Platform="Any CPU" -m
MSBuild version 17.4.1+9a89d02ff for .NET Framework
MSBUILD : error MSB1008: Only one project can be specified.
    Full command line: 'C:\Apps\MSBuild\vc\Current\Bin\MSBuild.exe  src/KeePass.sln -target:keepass2android-app -p:AndroidSdkDirectory="C:\Users\me\AppData\Local\Android\Sdk\" -p:BuildProjectReferences=true  -p:Platform="Any CPU" -m'
  Switches appended by response files:
'' came from 'C:\Apps\MSBuild\vc\Current\Bin\MSBuild.rsp'
Switch: CPU -m

For switch syntax, type "MSBuild -help"
make: *** [msbuild] Error 1
```
If I manually run `MSBuild` without those double-quotes, we move to yet another error:

**Problem 3: MSBuild missing/wrong location of `Xamarin.Android.CSharp.targets`**
```
C:\Users\me\source\forks\keepass2android>C:\Apps\MSBuild\vc\Current\Bin\MSBuild.exe  src/KeePass.sln -target:keepass2android-app -p:AndroidSdkDirectory=C:\Users\me\AppData\Local\Android\Sdk\ -p:BuildProjectReferences=true  -p:Platform="Any CPU" -m
MSBuild version 17.4.1+9a89d02ff for .NET Framework
Build started 2/6/2023 7:54:05 AM.
     1>Project "C:\Users\me\source\forks\keepass2android\src\KeePass.sln" on node 1 (keepass2android-app target(s)).
     1>ValidateSolutionConfiguration:
         Building solution configuration "Debug|Any CPU".
     1>Project "C:\Users\me\source\forks\keepass2android\src\KeePass.sln" (1) is building "C:\Users\me\source\forks\keepass2android\src\keepass2android\keepass2android-app.csproj" (2)
        on node 1 (default targets).
     2>C:\Users\me\source\forks\keepass2android\src\keepass2android\keepass2android-app.csproj(1978,11): error MSB4226: The imported project "C:\Apps\MSBuild\vc\Current\Bin\Xamarin\Android\Xamarin.Android.CSharp.targets" was not found. Also, tried to find "Xamarin\Android\Xamarin.Android.CSharp.targets" in the fallback search path(s) for $(MSBuildExtensionsPath) - "
       C:\Program Files (x86)\MSBuild" . These search paths are defined in "C:\Apps\MSBuild\vc\Current\Bin\MSBuild.exe.Config". Confirm that the path in the <Import> declaration is correct,
        and that the file exists on disk in one of the search paths.
     2>Done Building Project "C:\Users\me\source\forks\keepass2android\src\keepass2android\keepass2android-app.csproj" (default targets) -- FAILED.
     1>Done Building Project "C:\Users\me\source\forks\keepass2android\src\KeePass.sln" (keepass2android-app target(s)) -- FAILED.

Build FAILED.

       "C:\Users\me\source\forks\keepass2android\src\KeePass.sln" (keepass2android-app target) (1) ->
       "C:\Users\me\source\forks\keepass2android\src\keepass2android\keepass2android-app.csproj" (default target) (2) ->
         C:\Users\me\source\forks\keepass2android\src\keepass2android\keepass2android-app.csproj(1978,11): error MSB4226: The imported project "C:\Apps\MSBuild\vc\Current\Bin\Xamarin\Android\Xamarin.Android.CSharp.targets" was not found. Also, tried to find "Xamarin\Android\Xamarin.Android.CSharp.targets" in the fallback search path(s) for $(MSBuildExtensionsPath) -
        "C:\Program Files (x86)\MSBuild" . These search paths are defined in "C:\Apps\MSBuild\vc\Current\Bin\MSBuild.exe.Config". Confirm that the path in the <Import> declaration is correct, and that the file exists on disk in one of the search paths.

    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:00.05
```
And indeed, `C:\Apps\MSBuild\vc\Current\Bin\Xamarin\Android\Xamarin.Android.CSharp.targets` does not exist. It exists in `C:\Apps\MSBuild\vc\Xamarin\Android`. So now I've `mklink` hacked *that* directory, and the error changes again:

**Problem 4: Reference assemblies for MonoAndroid,Version=v1.0 were not found**
```
C:\Users\me\source\forks\keepass2android>C:\Apps\MSBuild\vc\Current\Bin\MSBuild.exe  src/KeePass.sln -target:keepass2android-app -p:AndroidSdkDirectory=C:\Users\me\AppData\Local\Android\Sdk\ -p:BuildProjectReferences=true  -p:Platform="Any CPU" -m
MSBuild version 17.4.1+9a89d02ff for .NET Framework
Build started 2/6/2023 8:05:31 AM.
     1>Project "C:\Users\me\source\forks\keepass2android\src\KeePass.sln" on node 1 (keepass2android-app target(s)).
     1>ValidateSolutionConfiguration:
         Building solution configuration "Debug|Any CPU".
     1>Project "C:\Users\me\source\forks\keepass2android\src\KeePass.sln" (1) is building "C:\Users\me\source\forks\keepass2android\src\keepass2android\keepass2android-app.csproj" (2)
        on node 1 (default targets).
     2>showFlavor:
         building flavor NoNet
     2>C:\Apps\MSBuild\vc\Current\Bin\Xamarin\Android\Xamarin.Android.Legacy.targets(234,5): error MSB3644: The reference assemblies for MonoAndroid,Version=v1.0 were not found. To resolve
       this, install the Developer Pack (SDK/Targeting Pack) for this framework version or retarget your application. You can download .NET Framework Developer Packs at https://aka.ms/msbui
       ld/developerpacks [C:\Users\me\source\forks\keepass2android\src\keepass2android\keepass2android-app.csproj]
     2>Done Building Project "C:\Users\me\source\forks\keepass2android\src\keepass2android\keepass2android-app.csproj" (default targets) -- FAILED.
     1>Done Building Project "C:\Users\me\source\forks\keepass2android\src\KeePass.sln" (keepass2android-app target(s)) -- FAILED.

Build FAILED.

       "C:\Users\me\source\forks\keepass2android\src\KeePass.sln" (keepass2android-app target) (1) ->
       "C:\Users\me\source\forks\keepass2android\src\keepass2android\keepass2android-app.csproj" (default target) (2) ->
       (_GetReferenceAssemblyPaths target) ->
         C:\Apps\MSBuild\vc\Current\Bin\Xamarin\Android\Xamarin.Android.Legacy.targets(234,5): error MSB3644: The reference assemblies for MonoAndroid,Version=v1.0 were not found. To resolve this, install the Developer Pack (SDK/Targeting Pack) for this framework version or retarget your application. You can download .NET Framework Developer Packs at https://aka.ms/msbuild/developerpacks [C:\Users\me\source\forks\keepass2android\src\keepass2android\keepass2android-app.csproj]

    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:00.12
```
I have `.NET 7.10.102 (x64) from Visual Studio` installed already, which is the latest/recommended SDK. At this point I feel like I'm out of options...

What versions of VS, Xamarin.Android, Android NDK, .NET SDK, MSBuild, etc do you folks use? Do you build on command line? I've now attempted this setup several times and I clearly am missing some secret sauce.

**Comment by hyproman** [comment]
It's probably worth noting that `MSBuild` version being used is now `v17.4.1` (that shipped with `2022 Build Tools`).
I also am now using a separate install of `GNU Make for Windows` binary `v3.81` instead of the one bundled with the `Android NDK`.

**Comment by tenzap** [comment]
The github actions jobs build with the makefile.


The windows build uses` windows-2022 `which has these tools installed https://github.com/actions/runner-images/blob/main/images/win/Windows2022-Readme.md

This is the workflow file which show how the build is done:
https://github.com/PhilippC/keepass2android/blob/master/.github/workflows/build.yml#L245

You may also want to have a look at the logs of the github actions jobs:
https://github.com/PhilippC/keepass2android/actions/runs/4068342552/jobs/7006779446

Are you launching `make` from PowerShell of from cmd?

You may try removing MSBuild from your PATH and let Makefile suggest you to use  vcvarsall.bat
See https://github.com/PhilippC/keepass2android/blob/master/Makefile#L66
and https://github.com/PhilippC/keepass2android/blob/master/Makefile#L86

**Comment by hyproman** [comment]
I appreciate the info, and apologies for the wait.

**TL;DR:** I was finally able to get a working build environment, albeit with one archaic MSBUILD path hack. The auto-detect logic in the `Makefile` doesn't quite work right with my version of `make` or environment, but I'll certainly live with it. Thank you for the help!

**-- details below --**

I have been attempting to build via `cmd` (I prefer linux bash, but I never spent enough time with PowerShell to be useful). Using `GIT Bash` or `GIT sh` also do work now, but all three methods require my MSBUILD path hack on my machine.

It's worth noting I was missing one of the C++ workflows in Visual Studio; I believe this may have been the root cause of  `Reference assemblies for MonoAndroid,Version=v1.0 were not found` errors.

The double-quoting issue around `AndroidSdkDirectory` was caused by my `ANDROID_*` path variables containing trailing backslashes. These aren't escaped properly, which causes the closing double-quote to become escaped, which causes `MSBuild` to parse the rest of the parameters incorrectly. Worth noting that `Build.readme.md` examples include these trailing backslashes.

To get `make` to run `MSBuild.exe` properly, I ended up having to explicitly set MSBUILD to the old 8.3 path notation. No other method worked for me.

"Pretty" Location: `C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Current\Bin\MSBuild.exe`
Short Version:  `C:\PROGRA~1\MICROS~2\2022\COMMUN~1\MSBuild\Current\Bin\MSBuild.exe`

Putting it all together:
```
set ANDROID_HOME=C:\Apps\Android\Sdk
set ANDROID_SDK_ROOT=%ANDROID_HOME%
set ANDROID_NDK_ROOT=%ANDROID_HOME%\ndk\25.2.9519653
set MSBUILD="C:\PROGRA~1\MICROS~2\2022\COMMUN~1\MSBuild\Current\Bin\MSBuild.exe"

make MSBUILD=%MSBUILD% %*
```
